### PR TITLE
databricks client: use exceptions and add logging, SparkUI requests control

### PR DESCRIPTION
**Pull request by guybortnikov**
This pull request improves databricks metadata client, adds more logging, and adds the ability to control its defaults.
Also, it adds support to control parameters when accessing SparkUI for metrics.

### CODE REVIEW

File: granulate_utils/metrics/spark.py [hunk 1]
```diff
@@ -268,7 +283,12 @@ def _spark_executor_metrics(self, running_apps: Dict[str, Tuple[str, str]]) -> I
         for app_id, (app_name, tracking_url) in running_apps.items():
             try:
                 base_url = get_request_url(self.master_address, tracking_url)
-                executors = rest_request_to_json(base_url, SPARK_APPS_PATH, app_id, "executors")
+                executors = rest_request_to_json(
+                    base_url,
+                    SPARK_APPS_PATH,
+                    app_id,
+                    "executors",
+                )
```
Comment (d3dave) [inline granulate_utils/metrics/spark.py:291]: revert this

File: granulate_utils/metadata/databricks_client.py [hunk 2]
```diff
@@ -46,12 +55,25 @@
...
+        self,
+        logger: logging.LoggerAdapter,
+        enable_retries: bool = True,
+        http_request_timeout: int = DEFAULT_REQUEST_TIMEOUT,
+        raise_on_failure: bool = False,
+        verify_request_ssl: bool = DEFAULT_REQUEST_VERIFY_SSL,
+    ) -> None:
         """
         When `enable_retries` is True, the wrapper will retry the request to the webui until it succeeds or until
         """
```
Comment (d3dave) [inline granulate_utils/metadata/databricks_client.py:68]: The sentence here is cut off 😕

File: granulate_utils/metadata/databricks_client.py [hunk 3]
```diff
@@ -85,24 +115,25 @@ def extract_relevant_metadata(self) -> Optional[Dict[str, str]]:
...
-        self.logger.info("Databricks get DBX environment metadata timeout, continuing...")
+
+            # No environment metadata yet, retry.
+            time.sleep(RETRY_INTERVAL_S)
+
+        self.logger.info("Databricks get DBX environment metadata timeout reached")
+
+        if self.raise_on_failure:
+            raise DatabricksMetadataFetchException(
+                "Failed to get DBX environment metadata in timeout " f"of {DATABRICKS_JOBNAME_TIMEOUT_S} seconds"
```
Comment (d3dave) [inline granulate_utils/metadata/databricks_client.py:135]:
```suggestion
                f"Timed out getting DBX environment metadata after {DATABRICKS_JOBNAME_TIMEOUT_S}s"
```

File: granulate_utils/metadata/databricks_client.py [hunk 4]
```diff
@@ -85,24 +115,25 @@ def extract_relevant_metadata(self) -> Optional[Dict[str, str]]:
...
+                    raise e
+
             if not self.enable_retries:
                 break
-        self.logger.info("Databricks get DBX environment metadata timeout, continuing...")
+
+            # No environment metadata yet, retry.
+            time.sleep(RETRY_INTERVAL_S)
+
+        self.logger.info("Databricks get DBX environment metadata timeout reached")
```
Comment (d3dave) [inline granulate_utils/metadata/databricks_client.py:131]: This message looks redundant if we are raising the exception below. Maybe we should only log this if we do not raise?

File: granulate_utils/metadata/databricks_client.py [hunk 5]
```diff
@@ -115,109 +146,107 @@ def _discover_apps_url(self) -> bool:
...
         """
-        Returns `includes spark.databricks.clusterUsageTags.clusterAllTags` tags as `Dict`.
+        Returns `includes spark.databricks.clusterUsageTags.clusterAllTags` tags as `dict`.
         In any case this function returns `None`, a retry is required.
         """
         if not os.path.isfile(DATABRICKS_METRICS_PROP_PATH):
+            self.logger.warning(f"{DATABRICKS_METRICS_PROP_PATH} was not found.")
             # We want to retry in case the cluster is still initializing, and the file is not yet deployed.
-            return None
+            raise SparkAppsURLDiscoveryException(f"{DATABRICKS_METRICS_PROP_PATH} was not found.")
```
Comment (d3dave) [inline granulate_utils/metadata/databricks_client.py:181]: Logging messages before exceptions looks redundant. All the exceptions are always logged by the caller anyway afaict. If the log level is important, then we can add a value to the exception classes that says which level they should be logged at.

**Review by d3dave — CHANGES_REQUESTED** [review]

File: granulate_utils/metadata/databricks_client.py (hunk 2 above)
Comment (d3dave) [reply granulate_utils/metadata/databricks_client.py:68]: Also needs doc-ing for the new values.

File: granulate_utils/metrics/spark.py (hunk 1 above)
Comment (guybortnikov) [reply granulate_utils/metrics/spark.py:291]: This is the git pre-commit hook, why revert? It's just alignment

File: granulate_utils/metadata/databricks_client.py (hunk 2 above)
Comment (guybortnikov) [reply granulate_utils/metadata/databricks_client.py:68]: Old @roi-granulate's code, removing the comment

File: granulate_utils/metadata/databricks_client.py (hunk 5 above)
Comment (guybortnikov) [reply granulate_utils/metadata/databricks_client.py:181]: I agree but I'm afraid of a situation where I'm unaware of all the callers, this is why I also added an argument indicating whether to raise or not.  
It's either:
a. We drop all the log messages
b. We'll log only if not raise arg is opted in

File: granulate_utils/metrics/spark.py (hunk 1 above)
Comment (d3dave) [reply granulate_utils/metrics/spark.py:291]: because it looks worse than the alternative

**Review by d3dave — APPROVED** [review]

File: granulate_utils/metrics/spark.py (hunk 1 above)
Comment (Jongy) [reply granulate_utils/metrics/spark.py:291]: @d3dave agree but it's the autoformatter, I'm not blocking @guybortnikov on this, we can fix the format later.

**Pull request merged**